qadence_libs/qinfo_tools/qng.py

```diff
@@ -0,0 +1,132 @@
+from __future__ import annotations
+
+from typing import Callable
+
+import torch
+from qadence import QuantumCircuit
+from torch.optim.optimizer import Optimizer, required
+
+from qadence_libs.qinfo_tools.qfi import get_quantum_fisher, get_quantum_fisher_spsa
+
+
+class QNG(Optimizer):
+    """Implements the Quantum Natural Gradient Algorithm."""
+
+    def __init__(
+        self,
+        params: tuple | torch.Tensor,
+        circuit: QuantumCircuit = required,
+        lr: float = required,
+        beta: float = 10e-3,
+    ):
+        defaults = dict(circuit=circuit, lr=lr, beta=beta)
+        super(QNG, self).__init__(params, defaults)
+
+    def __setstate__(self, state):  # type: ignore
+        super().__setstate__(state)
+
+    def step(self, closure: Callable | None = None) -> torch.Tensor:
+        """Performs a single optimization step of the QNG algorithm.
+
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            loss = closure()
+
+        for group in self.param_groups:
+            grad_vec = torch.Tensor([v.grad.data for v in group["params"] if v.requires_grad])
+
+            # Calculate the EXACT metric tensor
+            metric_tensor = (1 / 4) * get_quantum_fisher(
+                group["circuit"],
+                vparams_values=group["params"],
+            )
+
+            with torch.no_grad():
+                # Finite shift the metric tensor to avoid problems when inverting
+                metric_tensor = metric_tensor + group["beta"] * torch.eye(len(grad_vec))
+
+                # Get transformed gradient vector
+                metric_tensor_inv = torch.linalg.inv(metric_tensor)
+                transf_grad = torch.matmul(metric_tensor_inv, grad_vec)
+
+                # Update parameters
+                for i, p in enumerate(group["params"]):
+                    if p.grad is None:
+                        continue
+                    p.data.add_(transf_grad[i], alpha=-group["lr"])
+
+        return loss
+
+
+class QNG_SPSA(Optimizer):
```

Comment:
Couldn't you make that one a derived class from `QNG` ?

Reply:
Not sure about this. There is not much to inherit from `QNG` and I believe it's cleaner to separate both. Also, this is how it is usually implemented in Pytorch, just directly inheriting from the `Optimizer` class (see [Adam](https://pytorch.org/docs/stable/_modules/torch/optim/adam.html#Adam) and [AdamW](https://pytorch.org/docs/stable/_modules/torch/optim/adamw.html#AdamW) implementations in Pytorch)

Reply:
I kinda disagree: the initialiser is almost identical, `__setstate__` is common, only `step` differs but there are commonalities. I think you can design an abstract class (possibly a dataclass) with an extra argument to select either the standard `QNG` or the `SPSA` variant.